### PR TITLE
[boost-modular-build-helper] [boost-uninstall] use system layout

### PR DESCRIPTION
**Pull request by vitalyster**
**Describe the pull request**
Boost layout switched to `system`
- #### What does your PR fix?
  Boost build system provides different "layouts" to control the naming of the resulting libraries, and the locations of the installed files. Also it have "auto-linking" feature which selects correct library depending on Windows C++ compiler and its options.  For some [reason](https://github.com/microsoft/vcpkg/issues/694#issuecomment-281235155) Vcpkg decides to switch off auto-linking feature but still uses "versioned" layout. It creates some kind of "auto-linking" mechanism using CMake, but this breaks possibility to use libraries from other build systems. 
  This PR switched the build to "system" layout (which is already used for non-Windows platforms) and removes additional complexity added by vcpkg to link Boost libraries on Windows platform.

- #### Which triplets are supported/not supported? Have you updated the [CI baseline](https://github.com/microsoft/vcpkg/blob/master/scripts/ci.baseline.txt)?  
  all

- #### Does your PR follow the [maintainer guide](https://github.com/microsoft/vcpkg/blob/master/docs/maintainers/maintainer-guide.md)?  
  Yes

- #### If you have added/updated a port: Have you run `./vcpkg x-add-version --all` and committed the result?  
  Yes

### CODE REVIEW

**Review by JonLiu1993 — DISMISSED** [review]

**Comment by ras0219-msft** [comment]
The reason we do not use the system layout on Windows is because it is incompatible with Multi-Configuration CMake Generators like MSBuild while using the built-in `FindBoost.cmake` module.

The FindBoost module needs to be able to find both the release and debug library files for Multi-Config Generators to work, but the system layout does not have any naming difference between them. This is why our current approach is to "pin" all the relevant values (compiler, etc) such that FindBoost is able to find them while also being able to distinguish between Release and Debug.

Can you please go into detail about the precise problem this PR resolves including repro steps? Thanks for the PR!

**Comment by vitalyster** [comment]
1.
> find both the release and debug library files for Multi-Config Generators 

Everything works just fine, tested with minimal CMake project: Release directory contains release version of boost dlls and Debug directory contains Debug dlls, no need to have these creepy suffixes.

2.
>  precise problem this PR resolves

This PR resolves problem with other build systems: without system layout each build system should implement giant FindBoost.cmake analog for no reason.

**Comment by vitalyster** [comment]
3. Most of ports worked perfectly without suffixes, some ports like Boost have that only for legacy reasons which are not actual for vcpkg in my opinion.

**Comment by Neumann-A** [comment]
@vitalyster You don't understand the problem. Multi Config cmake generators need the suffix to be able to correctly link against the libraries and generate correctly setup `<package>-target.cmake` files.  If it is the same name CMake can only find one version of the libraries on its own.

**Comment by vitalyster** [comment]
@Neumann-A why it works for me and for thousands of other ports?

**Comment by Neumann-A** [comment]
> why it works for me and for thousands of other ports?

Because you don't know where the issues are and don't understand the multi config issues /  `<package>-target.cmake` generation issue. If you only ever use single configuration stuff and consumers link against the boost target instead of the `_LIBRARY` vars you probably never see these issue. But those issues exists and are annoying. At least the `d.lib` must stay for cmake to do the right thing. (I really don't care about the rest of the suffix)

**Comment by vitalyster** [comment]
@Neumann-A I tested a multi-config CMake build and it works. If you know some issues that can happen - please provide an example.

**Comment by vitalyster** [comment]
> At least the d.lib

I see the only reason to have a suffix is when all libraries are stored in a single directory. But in the vcpkg world debug libraries are stored in `debug` directory and this enough to distinct debug and release binaries

**Comment by Neumann-A** [comment]
> I tested a multi-config CMake build and it works.

probably only tested debug config.....

> I see the only reason to have a suffix is when all libraries are stored in a single directory.

wrong.. not the only reason. The reason to have `d` suffix is so that cmake can distinguish `_LIBRARY_DEBUG` from `_LIBRARY_RELEASE` and actually find the correct `DEBUG` and `RELEASE` version to setup the targets. Otherwise a `vcpkg-cmake-wrapper.cmake` is necessary to fix the targets/variables

> If you know some issues that can happen - please provide an example.

feel free to read #6045 and all associated issues/pr. This should keep you busy for at least a week :P

**Comment by vitalyster** [comment]
> cmake can distinguish _LIBRARY_DEBUG from _LIBRARY_RELEASE

1. They are stored in totally different directories: `<triplet>/lib/library_name.dll` and `<triplet>/debug/lib/library_name.dll`. All information previously encoded in suffixes are now in file path.

> https://github.com/microsoft/vcpkg/issues/6045 and all associated issues/pr.

The reason of postfix issues is postfix presence at all, it is not needed.
Well, if I see the port which does not have debug suffixes - is it a bug?

**Comment by Neumann-A** [comment]
> They are stored in totally different directories:

You are missing the point. CMake does not have a **default** concept of "different directories" for libraries. 
So what will `find_library(somelib_RELEASE NAMES somelib)` and `find_library(somelib_DEBUG NAMES somelib)` find? 
If you don't get it after that please look into your cmake module directory and look at some `Find<Modules>.cmake`. It doesn't matter if the information is encoded in the directory for cmake since cmake knows nothing about that and there is no way to tell it to cmake in a general configuration dependent way. 

> Well, if I see the port which does not have debug suffixes - is it a bug?

no but it either has a `vcpkg-cmake-wrapper.cmake` if it has a cmake `Find<Module>.cmake` or it just has pkgconfig files which are ok but don't work correctly either in cmake multi config.

**Comment by vitalyster** [comment]
> find_library(somelib_RELEASE NAMES somelib) and find_library(somelib_DEBUG NAMES somelib) find?

Why I should do these strange things? `find_library(Boost COMPONENTS ...)` with my patch just works. And Visual Studio builds successfully both Release and Debug targets.

> it either has a vcpkg-cmake-wrapper.cmake

Great! So vcpkg-cmake-wrapper do the right thing and we don't need to have suffixes.

> pkgconfig files which are ok but don't work correctly either in cmake multi config

Yes, and that is the job of vcpkg to provide a wrapper for multi config cmake and take the information from library directory or suffix.
You are missing the point that these suffixes makes vcpkg builds unusable from other build systems. While I can provide a triplet directory root (either release or debug one) with clean names for any build system and it will find libraries without any specific changes. Yes it will not work for "multi config" builds, but it is not needed because there is no such concept in other build systems.

**Comment by ras0219-msft** [comment]
Let's get back to the core problem seeking a solution.

>This PR resolves problem with other build systems: without system layout each build system should implement giant FindBoost.cmake analog for no reason.

Which buildsystem are you specifically having an issue with? I'm not sure if it has changed in recent versions, but the prebuilt binaries provided by Boost for Windows all are built using the default layout and not the system layout.

**Comment by vitalyster** [comment]
@ras0219-msft prebuild Boost binaries are using auto linking by default and that is assumed by a lot of custom build tools. These suffixes were made for Boost auto linking feature.
Quoting your answer in https://github.com/microsoft/vcpkg/issues/694#issuecomment-281235155

> we want to embrace the common, cross-platform core that CMake defines. Therefore, when using CMake, we model the same experience (i.e. no-auto-linking) as on other platforms so you can minimize per-platform differences

Without suffixes I can use vcpkg binaries in custom build environment, and there are no any reason to keep suffixes when auto linking is disabled.

**Comment by JonLiu1993** [comment]
@ras0219-msft ,Could you please help review this pr?

**Comment by ras0219-msft** [comment]
Sorry for taking a long time to get back to this issue, but we've thought more about the problem and come to the following conclusions:

1. Relying on the publicly documented boost feature of autolinking (which is supported by the official boost distribution) makes sense and is something that vcpkg should support. If that will resolve your case, then I move that we retitle this issue to be "Re-enable boost autolinking with MSVC".
2. While it would be convenient for boost to officially have the same binary names on all platforms, the official prebuilt windows binaries do not use the system layout, so it doesn't seem that us providing the system layout will overlap with existing users. Therefore, we do not see moving to the system layout as worth the costs to CMake's Multi-config generators. (MSBuild users are fine either way via our MSBuild integration).

**Comment by vitalyster** [comment]
Yes, I'm ok with autolinking support, but that may require a lot more work

**Comment by JonLiu1993** [comment]
@vitalyster, please run command `./vcpkg x-add-version boost-modular-build-helper --overwrite-version`and commit again

**Comment by JonLiu1993** [comment]
@vitalyster, Can you please resolve the conflicts against master?

**Comment by JonLiu1993** [comment]
@vitalyster, Can you please resolve the conflicts against master?

**Review by JonLiu1993 — APPROVED** [review]

File: versions/b-/boost-modular-build-helper.json [hunk 1]
```diff
@@ -1,7 +1,7 @@
 {
   "versions": [
     {
-      "git-tree": "d47cef023f39ac9d5f3a043d35cba1a74710e1b3",
+      "git-tree": "6511d7732378e174323311c423b00c175dbf8902",
       "version": "1.81.0",
       "port-version": 3
```
Comment (JonLiu1993) [inline versions/b-/boost-modular-build-helper.json:6]: @vitalyst, We can't modify the historical `git-tree` here, because a new record should be added, please revert the changes here and re-run the command `./vcpkg x-add-version boost-modular-build-helper --overwrite-version`

**Review by JonLiu1993 — CHANGES_REQUESTED** [review]

File: ports/boost-modular-build-helper/boost-modular-build.cmake [hunk 2]
```diff
@@ -40,17 +40,8 @@ function(boost_modular_build)
...
-            set(BOOST_LIB_DEBUG_SUFFIX -vc140-mt-gd.lib)
-        elseif(VCPKG_PLATFORM_TOOLSET MATCHES "v120")
-            set(BOOST_LIB_RELEASE_SUFFIX -vc120-mt.lib)
-            set(BOOST_LIB_DEBUG_SUFFIX -vc120-mt-gd.lib)
-        else()
-            set(BOOST_LIB_RELEASE_SUFFIX .lib)
-            set(BOOST_LIB_DEBUG_SUFFIX d.lib)
-        endif()
+        set(BOOST_LIB_RELEASE_SUFFIX .lib)
+        set(BOOST_LIB_DEBUG_SUFFIX .lib)
```
Comment (Neumann-A) [inline ports/boost-modular-build-helper/boost-modular-build.cmake:44]: there is a lot of more code down below which tries to rename the libs. Since this uses `--layout=system` the renaming becomes obsolete and the code for it should be removed. (but i doubt this will land anyways.)

Comment (Neumann-A) [reply ports/boost-modular-build-helper/boost-modular-build.cmake:44]: Also FindBoost.cmake will be broken with this change for multi config generators.

Comment (vitalyster) [reply ports/boost-modular-build-helper/boost-modular-build.cmake:46]: `vcpkg-cmake-wrapper.cmake` now set correct variables to help `FindBoost.cmake` which works in all configurations including `*-static` and `*-static-md`

Comment (Neumann-A) [reply ports/boost-modular-build-helper/boost-modular-build.cmake:46]: No it doesn't. Multi config is still broken. You cannot fix `find_library` by not touching `_RELEASE|_DEBUG` variables. As a reminder: The only way for CMake to differentiate between `_RELEASE|_DEBUG` variables is by filename. Since this PR removes the difference between configurations you need to put in a lot more work into the wrapper to make it work.

File: ports/boost-uninstall/vcpkg-cmake-wrapper.cmake [hunk 3]
```diff
@@ -1,11 +1,18 @@
-set(Boost_USE_STATIC_LIBS OFF)
+if (VCPKG_LIBRARY_LINKAGE STREQUAL dynamic)
```
Comment (dg0yt) [inline ports/boost-uninstall/vcpkg-cmake-wrapper.cmake:1]: Wrong time of check. This is evaluated when the package is used. But it must match the configuration at the time the port is built. i.e. it must be fixed at port build time.

File: ports/boost-modular-build-helper/boost-modular-build.cmake (hunk 2 above)
Comment (vitalyster) [reply ports/boost-modular-build-helper/boost-modular-build.cmake:46]: @Neumann-A [Tests](https://dev.azure.com/vcpkg/c1ee48cb-0df2-4ab3-8384-b1df5a79fe53/_build/results?buildId=86113&view=ms.vss-test-web.build-test-results-tab)

Failed: 0 (0.00%)
Passed: 4,146 (100.00%)
Other: 0 (0.00%)
Total: 4,146

Do you mean your testing system is broken?

Comment (Neumann-A) [reply ports/boost-modular-build-helper/boost-modular-build.cmake:46]: > Do you mean your testing system is broken?

Multi config usage is not tested by CI. CI just uses single config usage which is fine.

Comment (vitalyster) [reply ports/boost-modular-build-helper/boost-modular-build.cmake:46]: I test it using "vcpkg install boost" and got debug (in target/debug/ directory) and release libraries. 
I tested it using "vcpkg install cpp-netlib" and got debug and release libraries, correctly linked to debug and release boost libraries.

Comment (Neumann-A) [reply ports/boost-modular-build-helper/boost-modular-build.cmake:46]: > I tested it using "vcpkg install cpp-netlib" and got debug and release libraries, correctly linked to debug and release boost libraries.

That is single config usage. Try Ninja Multi Config or VS Generators and switch in the IDE between configurations.

Comment (vitalyster) [reply ports/boost-modular-build-helper/boost-modular-build.cmake:46]: Still no idea what a mystic issue you have there, I'm switching configurations, triplets, static libraries, dynamic CRT - everything works and linked to correct libraries
image

Comment (Neumann-A) [reply ports/boost-modular-build-helper/boost-modular-build.cmake:46]: I am talking about 
https://gitlab.kitware.com/cmake/cmake/-/blob/master/Modules/FindBoost.cmake#L2269
https://gitlab.kitware.com/cmake/cmake/-/blob/master/Modules/FindBoost.cmake#L2324
which will not have the correct values if you use a multi config generator using this PR. With this PR the values for both of the variables will be the same depending on how vcpkg has added itself to `CMAKE_PREFIX_PATH`. This breaks mutli config since you are either in Debug configuration and always linking the Release version or you are in Release and always linking the Debug lib. 
(what an uphill battle....if you still don't understand I don't know how to get more explicit)

Comment (vitalyster) [reply ports/boost-modular-build-helper/boost-modular-build.cmake:46]: You can simply provide a test case if you say 4146 tests in CI does not catch this issue. What is a purpose of CI there then?

Comment (Neumann-A) [reply ports/boost-modular-build-helper/boost-modular-build.cmake:46]: > CI does not catch this issue

You are overvaluing CI. CI mainly catches build regressions but not as many usage regressions. (Also CI only tests that the build works it doesn't test if whatever has been build actually works if run)

One last straw. Look at
https://github.com/microsoft/vcpkg/blob/master/ports/zlib/vcpkg-cmake-wrapper.cmake
and try to understand why it exists and then apply that knowledge to this PR (It is a similar issue but not exactly the same since this is more about multi config usage than wrongly generated cmake configs but the root cause is similar enough).
If you don't understand the issue after that you can basically close this PR since if you are not going to fix the fallout nobody will and in the current state this PR is not mergeable since VS multi config usage is a topic vcpkg cares about.  

> You can simply provide a test case

The test case is: Simple CMake project using and linking one of the boost libs; CMake Generator = -G Visual Studio 17 2022 -> Open generated project -> look at the Linker->Inputs->additional libraries/dependencies properties -> switch configurations in the IDE and observe that the additional libraries properties doesn't change (which is wrong.)

Comment (vitalyster) [reply ports/boost-modular-build-helper/boost-modular-build.cmake:46]: >  VS multi config usage is a topic vcpkg cares about.

How it can be possible to "care about" something without a single test case in CI?

> vcpkg-cmake-wrapper.cmake

This PR already have changes to vcpkg-cmake-wrapper.cmake in [boost-uninstall] port, thanks to useful comments from other reviewers.

>  why it exists

In case of `zlib` port it exists because of zlib build system provides different names (`d` suffix for debug builds). In case of boost and its system layout we not needed such changes and build will select correct library from triplet/name or triplet/debug/name.

Comment (Neumann-A) [reply ports/boost-modular-build-helper/boost-modular-build.cmake:46]: > In case of zlib port it exists because of zlib build system provides different names (d suffix for debug builds). In case of boost and its system layout we not needed such changes and build will select correct library from triplet/name or triplet/debug/name.

Yeah you really don't understand the issue and didn't take the time to read the blame PR describing what was going wrong. So the problem wasn't the debug suffix the problem was that zlib is called `libz` on linux/macosx independent of configuration which is why the wrapper is necessary to point cmake to the correct libraries in vcpkg. 
And what exactly is this PR doing ?!?! (Hmm probably exactly that, removing name differences.....)

Maybe it is time for others to tell you what is going on because I have the feeling you don't believe me.

Comment (vitalyster) [reply ports/boost-modular-build-helper/boost-modular-build.cmake:46]: Sure, I believe you, but I'm pretty sure this project should not rely on some magic knowledge of Ancients and should have public test cases for things it is "cared about". I'm happy to add such test cases when I will able to reproduce the issue which details you does not want to share with others.

**Comment by dg0yt** [comment]
FTR, there is a draft PR to get a little bit of multi-config usage covered, #27408. It doesn't include boost.
In any case, multi-config support shouldn't degrade with new PRs. After all, that's the whole point why vcpkg builds release and debug by default.

**Comment by JonLiu1993** [comment]
I will be converting your PR to draft status. The above suggested changes are only recommendations. If you are willing to adopt them, please click "ready for review" after making the modifications. If you do not wish to make any changes, please click "ready for review" directly. That way, I can be aware that you've responded since you can't modify the tags.

**Comment by JonLiu1993** [comment]
@vitalyster, Could you pleasae fix the conflicting files?

**Comment by JonLiu1993** [comment]
@vitalyster, please change both `boost-uninstall/vcpkg.json` and `boost-modular-build-helper/vcpkg.json` files, then run command `./vcpkg x-add-version --all --overwrite-version` and commit again.
